source/common/config/grpc_mux_impl.cc

```diff
@@ -232,7 +237,15 @@ void GrpcMuxImpl::onReceiveMessage(std::unique_ptr<envoy::api::v2::DiscoveryResp
     error_detail->set_message(e.what());
   }
   api_state_[type_url].request_.set_response_nonce(message->nonce());
-  sendDiscoveryRequest(type_url);
+
+  // If update is rejected continuously, this may become too chatty. Having back-off here gives
+  // chance for Envoy and Management Server to recover.
+  if (api_state_[type_url].request_.has_error_detail()) {
+    api_state_[type_url].enableRejectedTimer();
+  } else {
+    api_state_[type_url].resetRejectedState();
+    sendDiscoveryRequest(type_url);
```

Comment:
Might we double send if the timer has already fired in the epoll loop and queued the lambda with the `sendDiscoveryRequest()`?

Reply:
Yeah. There is a remote possibility of that but in that case shouldn't the management server detect this staleness based on version, nonce and ignore one?

Reply:
Ah..I see what you are saying - Because of the shared state in api_state_ both requests may go with the same version, nonce which is a problem. Any thoughts on how we can handle?

Reply:
nm..locking is not required here as timer also fires on the main thread looks like. So handled this case. PTAL.

Reply:
Yeah, we could race on some of this state. I think what would be a lot clearer actually is to have a queue for responses that we always add to on each send attempt. We can rate limit the removal of messages from this queue. That way, we preserve the property that every message the management server sends has a corresponding response and that the response has the same order as the original messages. I'm not sure how strictly necessary these properties are, but they will make it much easier to debug when things go wrong.

Reply:
@htuch I thought about this queuing model. One problem I saw with that model is queue size and the memory it occupies if we get a burst of failure messages. So I thought this approach may work better. If you think that is not a problem or can be mitigated some how and it is worth trying, I can give it a try. LMK.

Reply:
Won't the queue naturally limit its growth, since we won't be replying back to the management server and hence are applying back pressure of a sort?

Reply:
Not necessarily in all cases and for all resource types. Especially for end point updates and cluster updates that are triggered by backend changes, management servers can just listen to the changes on those events and push multiple updates to Envoy (though in a sequential way) on the open stream with out waiting for Envoy's response as they events mostly async in nature. When patches are rolled out and instances come and go, we have seen many updates being pushed to Envoy. So in such cases, queue may be limiting. Listeners and routes which at least in our case won't come in burst and they wait for envoy response so queue may be ok there.

Reply:
@htuch Assuming that queue sizes won't grow abnormally (probably it may be a reasonable assumption/compromise), I have tried the queue based approach.
Here is the flow
- When we get response from Management Server, we put the request in Queue and fire a timer that immediately looks at the queue
- We check if we are under ratelimit
     - if yes call `sendDiscoveryRequest` immediately.
     - if we are out of rate limit, we just enable the timer to check little later when rate limits 
        would be possibly available.
So in normal cases every thing would be called immediately. This solution has the added benefit of rate limiting success calls as well with order maintained.
Here is the branch that has the implementation https://github.com/ramaraochavali/envoy/commit/065f107adbc30a8493f03d21538c3a2af06cf638
See if it makes sense. If this is what we are looking for, I will make tests pass with that and update the PR.

Reply:
Yes, this looks like what I was thinking, thanks. I think there are some technicalities around the timer callback to think about:
1. We should just have a single queue for all messages on the channel, so we preserve ordering across different API types (easier to debug and reason about).
2. The queue may grow to some extent, let's say we end up with a handful of requests queued up. We need to be able to drain this in a sane way. I propose that we keep draining it while we're within the rate limit quota.
3. When you are out of quota, we should delay until we have quota again. This is neither a fixed timer or a jittered exponential; can we have some way to obtain from the rate limiter a suggested sleep period?

Reply:
Ok. Great. 
I am fine with single queue across all APIs. On the timer sleep, I was also thinking the same that it should be derived from rate limiter's rate filling factor. I will update.
Since we have increased the scope of rate liming from rejected cases to all request etc, I will close this PR, open a new one with the changes so that we can discuss there.